jplag/src/main/java/de/jplag/JPlagComparison.java

```diff
@@ -1,103 +1,32 @@
+ * @param firstSubmission is the first of the two submissions.
+ * @param secondSubmission is the second of the two submissions.
+ * @param matches is the unmodifiable list of all matches between the two submissions.
  */
-public class JPlagComparison { // FIXME TS: contains a lot of code duplication
-
-    private static final int ROUNDING_FACTOR = 10;
-
-    private final Submission firstSubmission;
-    private final Submission secondSubmission;
-
-    private final List<Match> matches;
-
-    public JPlagComparison(Submission firstSubmission, Submission secondSubmission) {
-        this.firstSubmission = firstSubmission;
-        this.secondSubmission = secondSubmission;
-        matches = new ArrayList<>();
-    }
-
-    /**
-     * Add a match to the comparison (token indices and number of tokens), if it does not overlap with the existing matches.
-     * @see Match#Match(int, int, int)
-     */
-    /* package-private */ final void addMatch(int startOfFirst, int startOfSecond, int length) {
-        for (Match match : matches) {
-            if (match.overlap(startOfFirst, startOfSecond, length)) {
-                return;
-            }
-        }
-        matches.add(new Match(startOfFirst, startOfSecond, length));
-    }
-
-    @Override
-    public boolean equals(Object other) {
-        if (other == this) {
-            return true;
-        }
-        if (!(other instanceof JPlagComparison otherComparison)) {
-            return false;
-        }
-        return firstSubmission.equals(otherComparison.getFirstSubmission()) && secondSubmission.equals(otherComparison.getSecondSubmission())
-                && matches.equals(otherComparison.matches);
-    }
-
-    @Override
-    public int hashCode() {
-        return Objects.hash(firstSubmission, secondSubmission, matches);
-    }
-
-    /**
-     * @return the base code matches of the first submission.
-     */
-    public JPlagComparison getFirstBaseCodeMatches() {
-        return firstSubmission.getBaseCodeComparison();
-    }
-
+public record JPlagComparison(Submission firstSubmission, Submission secondSubmission, List<Match> matches) {
     /**
-     * @return the first of the two submissions.
+     * Initializes a new comparison.
+     * @param firstSubmission is the first of the two submissions.
+     * @param secondSubmission is the second of the two submissions.
+     * @param matches is the list of all matches between the two submissions.
      */
-    public Submission getFirstSubmission() {
-        return firstSubmission;
-    }
-
-    /**
-     * @return all matches between the two submissions.
-     */
-    public List<Match> getMatches() {
-        return matches;
+    public JPlagComparison(Submission firstSubmission, Submission secondSubmission, List<Match> matches) {
+        this.firstSubmission = firstSubmission;
+        this.secondSubmission = secondSubmission;
+        this.matches = Collections.unmodifiableList(matches);
```

Comment:
I do not like the use of an unmodifiable list here! Upon using this record, there is no indication that the list of matches is unmodifiable. If misused, a crash will only occur at runtime. This is a badly designed API, in my opinion.

Reply:
You could use List<? extends Match> but that does not guarantee that methods like remove(Object o) will be invoked.
Java has no real alternative. Always copying isn't the best option in my opinion and records should act as some kind of value type :)

Reply:
So what is your suggestion? 😅

Reply:
Either we override the getter, or we turn it into a class again. Either way, I am in favor of avoiding `Collections.unmodifiableList` as much as possible, especially if it concerns the public API. However, I am heavily biased here by my pet peeve regarding the Java devs disregard for Liskov! We can also discuss our policy regarding that internally today.

Reply:
Agreed in maintainer meeting to use unmodifiable lists with an according documentation. Thus this resolves without further action.